File: data structures and algorithms/python/08_028.py
import heapq
from collections import deque

class KNearestInWindow:
   def __init__(self, window_size, k):
      self.window_size = window_size
      self.k = k
      self.window = deque()
      # Max heap containing (distance, point) pairs
      self.heap = []
   
   def distance(self, point):
      # Euclidean distance to origin
      return point[0]**2 + point[1]**2
   
   def add_point(self, point):
      dist = self.distance(point)
      
      # Add point to window
      self.window.append((dist, point))
      
      # Remove oldest point if window is too large
      if len(self.window) > self.window_size:
         oldest_dist, oldest_point = self.window.popleft()
         
         # Check if oldest point was in heap
         in_heap = False
         for i, (d, p) in enumerate(self.heap):
            if p == oldest_point:
               in_heap = True
               break
         
         # If oldest point was in heap, rebuild the heap
         if in_heap:
            self.rebuild_heap()
         
      # Add new point to heap if closer than farthest
      # or if heap isn't full yet
      if len(self.heap) < self.k:
         heapq.heappush(self.heap, (-dist, point))
      elif -dist > self.heap[0][0]:
         heapq.heappushpop(self.heap, (-dist, point))
   
   def rebuild_heap(self):
      # Get all points in current window
      points_in_window = [p for _, p in self.window]
      
      # Clear heap and rebuild
      self.heap = []
      for point in points_in_window:
         dist = self.distance(point)
         if len(self.heap) < self.k:
            heapq.heappush(self.heap, (-dist, point))
         elif -dist > self.heap[0][0]:
            heapq.heappushpop(self.heap, (-dist, point))
   
   def get_k_nearest(self):
      # Return the k nearest points
      return [point for _, point in sorted(
         self.heap, key=lambda x: -x[0])]

# Test with example
points = [(1, 2), (3, 4), (0, 1), (5, 2), 
         (2, 0), (1, 5), (3, 1)]
window_size = 5
k = 3

knn = KNearestInWindow(window_size, k)
for point in points:
   knn.add_point(point)

print(knn.get_k_nearest())  # [(2, 0), (0, 1), (3, 1)]

File: data structures and algorithms/cs/08_028.cs
using System;
using System.Collections.Generic;

class Point {
    public int X { get; set; }
    public int Y { get; set; }

    public Point(int x, int y) {
        X = x;
        Y = y;
    }

    public int DistanceToOrigin() {
        return X * X + Y * Y;
    }

    public override bool Equals(object obj) {
        if (this == obj) return true;
        if (!(obj is Point)) return false;
        Point other = (Point)obj;
        return X == other.X && Y == other.Y;
    }

    public override int GetHashCode() {
        return HashCode.Combine(X, Y);
    }

    public override string ToString() {
        return $"({X}, {Y})";
    }
}

class KNearestInWindow {
    private readonly int windowSize;
    private readonly int k;
    private readonly LinkedList<Point> window;
    private readonly PriorityQueue<Point> heap;

    public KNearestInWindow(int windowSize, int k) {
        this.windowSize = windowSize;
        this.k = k;
        this.window = new LinkedList<Point>();
        this.heap = new PriorityQueue<Point>((p1, p2) => p2.DistanceToOrigin().CompareTo(p1.DistanceToOrigin()));
    }

    public void AddPoint(Point point) {
        window.AddLast(point);

        if (window.Count > windowSize) {
            Point oldest = window.First.Value;
            window.RemoveFirst();

            if (heap.Contains(oldest)) {
                RebuildHeap();
            }
        }

        if (heap.Count < k) {
            heap.Enqueue(point);
        } else if (point.DistanceToOrigin() < heap.Peek().DistanceToOrigin()) {
            heap.Dequeue();
            heap.Enqueue(point);
        }
    }

    private void RebuildHeap() {
        heap.Clear();
        var points = new List<Point>(window);
        points.Sort((p1, p2) => p1.DistanceToOrigin().CompareTo(p2.DistanceToOrigin()));
        for (int i = 0; i < Math.Min(k, points.Count); i++) {
            heap.Enqueue(points[i]);
        }
    }

    public List<Point> GetKNearest() {
        var result = new List<Point>(heap);
        result.Sort((p1, p2) => p1.DistanceToOrigin().CompareTo(p2.DistanceToOrigin()));
        return result;
    }

    public static void Main(string[] args) {
        int[][] pointsArray = {
            new int[] { 1, 2 }, new int[] { 3, 4 }, new int[] { 0, 1 }, new int[] { 5, 2 },
            new int[] { 2, 0 }, new int[] { 1, 5 }, new int[] { 3, 1 }
        };
        int windowSize = 5;
        int k = 3;

        var knn = new KNearestInWindow(windowSize, k);
        foreach (var p in pointsArray) {
            knn.AddPoint(new Point(p[0], p[1]));
        }

        var result = knn.GetKNearest();
        Console.WriteLine("K nearest points: " + string.Join(", ", result));
    }
}